File: src/main/java/pro/tuscan/adapter/api/LatestProfilesEndpoint.java
package pro.tuscan.adapter.api;

import pro.tuscan.adapter.api.request.LatestProfileRequest;
import pro.tuscan.domain.profiles.LatestProfile;
import pro.tuscan.domain.profiles.LatestProfileService;
import io.micrometer.core.annotation.Timed;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.EmitterProcessor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.MediaType.TEXT_EVENT_STREAM_VALUE;

@RequestMapping("/tuscan-api/latest-profiles")
@RestController
@Timed("endpoint.latestProfiles")
class LatestProfilesEndpoint {

    private final EmitterProcessor<ServerSentEvent<LatestProfile>> emitter;
    private final LatestProfileService latestProfileService;

    LatestProfilesEndpoint(LatestProfileService latestProfileService) {
        this.emitter = EmitterProcessor.create(false);
        this.latestProfileService = latestProfileService;
    }

    @GetMapping
    public Flux<LatestProfile> getLatestProfiles() {
        return latestProfileService.findLatestProfiles();
    }

    @ResponseStatus(CREATED)
    @PostMapping
    public Mono<LatestProfile> saveProfile(@RequestBody LatestProfileRequest request) {

        return latestProfileService.save(request).doOnSuccess(profile -> {
            try {
                emitter.onNext(ServerSentEvent.builder(profile).build());
            } catch (Exception e) {
                emitter.onError(e);
            }
        });
    }

    @GetMapping(value = "/sse", produces = TEXT_EVENT_STREAM_VALUE)
    public Flux<ServerSentEvent<LatestProfile>> events() {
        return getEvents();
    }

    private Flux<ServerSentEvent<LatestProfile>> getEvents() {
        return emitter;
    }
}

File: src/main/java/pro/tuscan/adapter/api/LatestProfilesEndpoint.kt
package pro.tuscan.adapter.api

import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.MediaType
import org.springframework.http.codec.ServerSentEvent
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import pro.tuscan.domain.profiles.LatestProfile
import pro.tuscan.domain.profiles.LatestProfileService
import reactor.core.publisher.EmitterProcessor
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.math.BigDecimal

@RequestMapping("/tuscan-api/latest-profiles")
@RestController
class LatestProfilesEndpoint(private val latestProfileService: LatestProfileService) {

    private val emitter: EmitterProcessor<ServerSentEvent<LatestProfile>> = EmitterProcessor.create(false)

    @ResponseStatus(CREATED)
    @PostMapping
    fun saveProfile(@RequestBody request: LatestProfileRequest): Mono<LatestProfile> =
            latestProfileService.save(request).doOnSuccess { profile ->
                try {
                    emitter.onNext(ServerSentEvent.builder(profile).build())
                } catch (e: Exception) {
                    emitter.onError(e)
                }
            }

    @GetMapping
    fun getLatestProfiles(): Flux<LatestProfile> = latestProfileService.findLatestProfiles()

    @GetMapping(value = ["/sse"], produces = [MediaType.TEXT_EVENT_STREAM_VALUE])
    fun events(): Flux<ServerSentEvent<LatestProfile>> = emitter
}

data class LatestProfileRequest(val nickname: String,
                                val avatarUrl: String,
                                val level: Number,
                                val elo: Number,
                                val kdRatio: BigDecimal)
